File: src/lib/perf/ipcBench.ts
/**
 * JS half of the Rust↔JS IPC end-to-end bench — Perf Phase 0 criterion 3
 * (`specs/perf-p0-measurement-baseline-spec.md`; Rust half in
 * `src-tauri/src/ipcbench.rs`).
 *
 * Runs only when the REAL shell was launched in bench mode:
 *
 *   PHASR_IPC_BENCH=1 pnpm tauri dev
 *
 * `main.tsx` dynamically imports this in DEV builds; the first invoke
 * returns `null` in a normal launch (or throws in a plain browser) and
 * nothing else happens. In bench mode it drives the matrix — base64+JSON
 * `PtyEvent` vs raw bytes vs the shipping policy, at eval-path and
 * fetch-path sizes — through a real `Channel`, measuring on the ONE clock
 * that can see both ends of the hop:
 *
 *   - **one-shot**: invoke-start → the channel message's `onmessage`,
 *     minus the measured no-payload invoke RTT = net delivery cost of one
 *     chunk (reported both raw and net);
 *   - **stream**: 200 chunks sent back-to-back from Rust; first→last
 *     arrival span gives per-chunk pipeline cost and MB/s — the number
 *     that bounds flood.
 *
 * Each size is measured three ways: `json` (the pre-P4 wire), `raw` (bytes
 * with no envelope) and `auto` (**what ships** — `pty_stream::output_body`,
 * which picks between them by size). Run it on a RELEASE build: on a debug
 * build serde is roughly 30x slower than it ships, which flatters `raw`
 * enough to invert the conclusion.
 *
 * Results go back through `ipc_bench_report`, which prints `IPCBENCH`
 * lines on the launching terminal and exits the shell (unless
 * PHASR_IPC_BENCH=hold).
 */

import { Channel, invoke } from "@tauri-apps/api/core";

interface BenchCase {
  label: string;
  format: "json" | "raw" | "auto";
  /** Raw chunk bytes. JSON envelope ≈ 40 + ceil(size/3)*4. */
  size: number;
}

/**
 * Grouped by chunk size, so each size has a BEFORE (json), the pure-raw
 * alternative, and what ships — the comparison phase 4's criterion 2 asks
 * for, and the one an average would hide.
 *
 * Thresholds verified in tauri 2.11.2 `src/ipc/channel.rs`: JSON flips
 * eval→fetch at 8192 B of serialized JSON, raw at 1024 B of payload.
 *
 * The 4 KiB pair is the awkward one and is here on purpose: base64+JSON of
 * 4 KiB is ~5.5 KB and still takes `eval`, while 4 KiB of raw bytes is over
 * the 1024 B raw threshold and takes `fetch`. That is the only cell where
 * the raw move *changes transport path*, so it is the only cell where it
 * could plausibly lose.
 *
 * 512 B matters more since the leading-edge flush: a keystroke echo is now
 * its own small chunk rather than something coalesced into the next repaint.
 */
const CASES: BenchCase[] = [
  { label: "json   512B chunk (~723B JSON → eval)", format: "json", size: 512 },
  { label: "raw    512B chunk (raw → eval)", format: "raw", size: 512 },
  { label: "auto   512B chunk (SHIPPING)", format: "auto", size: 512 },
  { label: "json  4KiB chunk (~5.5KB JSON → eval)", format: "json", size: 4096 },
  { label: "raw   4KiB chunk (raw → fetch)", format: "raw", size: 4096 },
  { label: "auto  4KiB chunk (SHIPPING)", format: "auto", size: 4096 },
  { label: "json 32KiB chunk (~43.7KB JSON → fetch)", format: "json", size: 32768 },
  { label: "raw  32KiB chunk (raw → fetch)", format: "raw", size: 32768 },
  { label: "auto 32KiB chunk (SHIPPING)", format: "auto", size: 32768 },
];

const ONESHOT_REPS = 30;
const STREAM_COUNT = 200;

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

function p50(xs: number[]): number {
  const sorted = [...xs].sort((a, b) => a - b);
  return sorted[Math.floor(sorted.length / 2)] ?? 0;
}

async function oneShot(c: BenchCase): Promise<number> {
  const channel = new Channel<unknown>();
  let arrived: () => void;
  const arrival = new Promise<void>((resolve) => {
    arrived = resolve;
  });
  channel.onmessage = () => arrived();
  const t0 = performance.now();
  await invoke("ipc_bench_send", {
    channel,
    size: c.size,
    count: 1,
    format: c.format,
  });
  await arrival;
  return performance.now() - t0;
}

async function stream(
  c: BenchCase,
): Promise<{ spanMs: number; totalMs: number; rustMs: number }> {
  const channel = new Channel<unknown>();
  let first = 0;
  let seen = 0;
  let done: (spanMs: number) => void;
  const span = new Promise<number>((resolve) => {
    done = resolve;
  });
  channel.onmessage = () => {
    seen += 1;
    if (seen === 1) first = performance.now();
    if (seen === STREAM_COUNT) done(performance.now() - first);
  };
  const t0 = performance.now();
  const rustMs = await invoke<number>("ipc_bench_send", {
    channel,
    size: c.size,
    count: STREAM_COUNT,
    format: c.format,
  });
  const spanMs = await span;
  return { spanMs, totalMs: performance.now() - t0, rustMs };
}

export async function runIpcBenchIfRequested(): Promise<void> {
  let config: unknown;
  try {
    config = await invoke("ipc_bench_config");
  } catch {
    return; // not a Tauri shell (Playwright / plain browser)
  }
  if (!config) return; // real shell, not in bench mode

  // Let boot settle so the numbers are the transport's, not startup's.
  await sleep(2000);
  const lines: string[] = [];
  try {
    // No-payload invoke RTT — the baseline subtracted from one-shots.
    const noop: number[] = [];
    for (let i = 0; i < ONESHOT_REPS; i++) {
      const t0 = performance.now();
      await invoke("ipc_bench_config");
      noop.push(performance.now() - t0);
    }
    const noopP50 = p50(noop);
    lines.push(
      `invoke-noop RTT p50 ${noopP50.toFixed(2)}ms min ${Math.min(...noop).toFixed(2)}ms (n=${ONESHOT_REPS})`,
    );

    for (const c of CASES) {
      const rtts: number[] = [];
      for (let i = 0; i < ONESHOT_REPS; i++) rtts.push(await oneShot(c));
      const oneP50 = p50(rtts);
      const s = await stream(c);
      const perChunk = s.spanMs / (STREAM_COUNT - 1);
      const mbPerSec =
        (c.size * STREAM_COUNT) / 1_048_576 / (s.totalMs / 1000);
      lines.push(
        `${c.label}: one-shot p50 ${oneP50.toFixed(2)}ms (net ~${(oneP50 - noopP50).toFixed(2)}ms) ` +
          `min ${Math.min(...rtts).toFixed(2)}ms | stream x${STREAM_COUNT}: ` +
          `${perChunk.toFixed(3)}ms/chunk, ${mbPerSec.toFixed(1)}MB/s payload ` +
          `(total ${s.totalMs.toFixed(0)}ms, rust send-loop ${s.rustMs.toFixed(0)}ms)`,
      );
    }
  } catch (err) {
    lines.push(`FAILED: ${String(err)}`);
  }
  await invoke("ipc_bench_report", { lines });
}

File: src-tauri/src/ipcbench.rs
//! Rust↔JS end-to-end IPC bench — Perf Phase 0 criterion 3
//! (`specs/perf-p0-measurement-baseline-spec.md`).
//!
//! The mocked-IPC e2e harness synthesizes base64 on the JS side and can
//! never measure the real hop, so this harness runs **inside the real
//! shell**: launch `PHASR_IPC_BENCH=1 pnpm tauri dev` and the frontend
//! (`src/lib/perf/ipcBench.ts`, dev builds only) detects bench mode via
//! `ipc_bench_config`, drives the matrix through a real `tauri::ipc::Channel`
//! — the exact transport `PtyEvent` ships on — and reports back through
//! `ipc_bench_report`, which prints `IPCBENCH` lines and exits the app.
//!
//! What the matrix separates (verified against tauri 2.11.2
//! `src/ipc/channel.rs`):
//!
//! * **JSON < 8192 B** → delivered by `webview.eval` directly
//!   (`MAX_JSON_DIRECT_EXECUTE_THRESHOLD`). A 4 KiB PTY chunk base64s to
//!   ~5.5 KiB of JSON: the keystroke-echo path.
//! * **JSON ≥ 8192 B** → parked in the global
//!   `ChannelDataIpcQueue(Arc<Mutex<HashMap>>)` shared by every channel in
//!   the app — all PTYs included — then pulled by the webview via
//!   `plugin:__TAURI_CHANNEL__|fetch`. A 32 KiB coalescer chunk (~43.7 KiB
//!   JSON) always takes this path: the flood path.
//! * **Raw bytes** flip eval→fetch at 1024 B
//!   (`MAX_RAW_DIRECT_EXECUTE_THRESHOLD`) and skip base64+JSON entirely —
//!   what Phase 4 proposes.
//!
//! All timing happens on the JS clock (invoke-start → `onmessage`, and
//! first→last arrival for throughput), so no cross-clock arithmetic; the
//! Rust-side send-loop time is returned too, labeled as such.
//!
//! Nothing here is reachable in production: every command errors unless
//! `PHASR_IPC_BENCH` is set in the app's environment, and the frontend
//! runner is `import.meta.env.DEV`-gated on top.

use std::time::Instant;

use tauri::ipc::{Channel, InvokeResponseBody};
use tauri::AppHandle;

use crate::pty::PtyEvent;

fn enabled() -> bool {
    std::env::var_os("PHASR_IPC_BENCH").is_some()
}

/// `PHASR_IPC_BENCH=hold` keeps the app alive after the report, for
/// poking at the HUD in the same session; anything else exits.
fn exit_when_done() -> bool {
    std::env::var("PHASR_IPC_BENCH").map(|v| v != "hold").unwrap_or(true)
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IpcBenchConfig {
    exit_when_done: bool,
}

/// Escape-dense payload bytes, like agent TUI output (an SGR every 16
/// visible bytes). Content cannot change base64 size, but it keeps any
/// future non-base64 comparison honest.
fn corpus_bytes(n: usize) -> Vec<u8> {
    let mut out = Vec::with_capacity(n + 16);
    let mut i = 0usize;
    while out.len() < n {
        if i % 16 == 0 {
            out.extend_from_slice(b"\x1b[38;5;123m");
        }
        out.push(b'a' + (i % 26) as u8);
        i += 1;
    }
    out.truncate(n);
    out
}

/// The bytes the **pre-P4** path put on the wire for one chunk:
/// `PtyEvent::Output` through serde (base64 + JSON envelope).
///
/// Still built here, and only here: it is the BEFORE column. Since P4 the
/// shipping path is `InvokeResponseBody::Raw` of the chunk itself, which is
/// exactly what the `"raw"` format below sends — so one run of this bench
/// prints both sides of the change, on the same machine, through the same
/// real `Channel`.
fn json_body(chunk: &[u8]) -> Result<String, String> {
    serde_json::to_string(&PtyEvent::Output {
        task_id: "bench".into(),
        // Not serialized (backend-internal cursor), so it costs the wire
        // nothing — the envelope this measures is unchanged.
        log_offset: 0,
        chunk: bytes::Bytes::copy_from_slice(chunk),
    })
    .map_err(|e| e.to_string())
}

/// `None` unless the shell was launched in bench mode — the frontend's
/// cheap "should I run at all" probe, and the RTT no-op baseline.
#[tauri::command]
pub fn ipc_bench_config() -> Option<IpcBenchConfig> {
    enabled().then(|| IpcBenchConfig {
        exit_when_done: exit_when_done(),
    })
}

/// Send `count` chunks of `size` raw bytes through the channel in the
/// given wire format. Three of them:
///
/// * `"json"` — the pre-P4 wire: base64 inside a JSON `PtyEvent`.
/// * `"raw"` — bytes with no envelope, which P4 originally proposed for
///   everything.
/// * `"auto"` — **what actually ships**: `pty_stream::output_body`, which
///   keeps a chunk in the JSON envelope while that envelope still fits
///   tauri's `eval` threshold and goes raw above it. Measured through the
///   same function the forwarder calls, so the bench cannot drift from the
///   policy it is reporting on.
///
/// Returns the Rust-side send-loop wall time in ms — serialization plus
/// handing off to the webview transport; delivery is measured on the JS
/// side.
#[tauri::command]
pub fn ipc_bench_send(
    channel: Channel<InvokeResponseBody>,
    size: usize,
    count: usize,
    format: String,
) -> Result<f64, String> {
    if !enabled() {
        return Err("PHASR_IPC_BENCH is not set".into());
    }
    if size > 4 * 1024 * 1024 || count > 10_000 {
        return Err("bench bounds exceeded".into());
    }
    let bytes = corpus_bytes(size);
    let t0 = Instant::now();
    match format.as_str() {
        "json" => {
            for _ in 0..count {
                channel
                    .send(InvokeResponseBody::Json(json_body(&bytes)?))
                    .map_err(|e| e.to_string())?;
            }
        }
        "raw" => {
            for _ in 0..count {
                channel
                    .send(InvokeResponseBody::Raw(bytes.clone()))
                    .map_err(|e| e.to_string())?;
            }
        }
        "auto" => {
            for _ in 0..count {
                channel
                    .send(crate::commands::pty_stream::output_body(
                        "bench",
                        bytes::Bytes::copy_from_slice(&bytes),
                    ))
                    .map_err(|e| e.to_string())?;
            }
        }
        other => return Err(format!("unknown format `{other}`")),
    }
    Ok(t0.elapsed().as_secs_f64() * 1000.0)
}

/// The frontend's results land here so they end up on the terminal that
/// launched the shell (the webview console dies with the window).
#[tauri::command]
pub fn ipc_bench_report(app: AppHandle, lines: Vec<String>) -> Result<(), String> {
    if !enabled() {
        return Err("PHASR_IPC_BENCH is not set".into());
    }
    for line in &lines {
        eprintln!("IPCBENCH {line}");
    }
    eprintln!("IPCBENCH done");
    if exit_when_done() {
        std::thread::spawn(move || {
            // Let the invoke response reach the webview first.
            std::thread::sleep(std::time::Duration::from_millis(300));
            app.exit(0);
        });
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// The cargo-test half — what a process-local test can measure honestly
// ---------------------------------------------------------------------------

/// Rust-side component costs of the hop, gated exactly like `perfbench.rs`
/// (`PHASR_BENCH=1 cargo test --lib ipcbench -- --ignored --nocapture`).
/// A cargo test has no webview, so this measures serialization plus the
/// channel machinery up to the runtime boundary — the end-to-end numbers
/// come from the in-app run above. Numbers labeled accordingly in the
/// spec's Baseline table.
#[cfg(test)]
mod tests {
    use super::*;

    const REPS: usize = 200;

    fn median(mut xs: Vec<f64>) -> f64 {
        xs.sort_by(|a, b| a.partial_cmp(b).unwrap());
        if xs.is_empty() {
            return f64::NAN;
        }
        xs[xs.len() / 2]
    }

    #[test]
    #[ignore = "benchmark, not a gate — PHASR_BENCH=1 … -- --ignored --nocapture"]
    fn bench_ipc_rust_side_costs() {
        if std::env::var_os("PHASR_BENCH").is_none() {
            eprintln!("IPCBENCH skipped: set PHASR_BENCH=1");
            return;
        }
        for (label, size) in [("4KiB (eval path)", 4096usize), ("32KiB (fetch path)", 32768)] {
            let bytes = corpus_bytes(size);

            // Serialize-only: the base64+JSON envelope build.
            let mut ser_us = Vec::with_capacity(REPS);
            for _ in 0..REPS {
                let t = Instant::now();
                let body = json_body(&bytes).unwrap();
                ser_us.push(t.elapsed().as_secs_f64() * 1e6);
                std::hint::black_box(body);
            }

            // Serialize + channel send into a no-op sink: adds the channel
            // plumbing (`IpcResponse::body`, the boxed `on_message` call)
            // without a webview behind it.
            let sink: Channel<InvokeResponseBody> =
                Channel::new(|body| {
                    std::hint::black_box(&body);
                    Ok(())
                });
            let mut send_json_us = Vec::with_capacity(REPS);
            for _ in 0..REPS {
                let t = Instant::now();
                sink.send(InvokeResponseBody::Json(json_body(&bytes).unwrap()))
                    .unwrap();
                send_json_us.push(t.elapsed().as_secs_f64() * 1e6);
            }
            let mut send_raw_us = Vec::with_capacity(REPS);
            for _ in 0..REPS {
                let t = Instant::now();
                sink.send(InvokeResponseBody::Raw(bytes.clone())).unwrap();
                send_raw_us.push(t.elapsed().as_secs_f64() * 1e6);
            }

            eprintln!(
                "IPCBENCH rust-side {label}: serialize {:.1}us  send(json) {:.1}us  send(raw) {:.1}us  (median of {REPS}; no webview behind the sink)",
                median(ser_us),
                median(send_json_us),
                median(send_raw_us),
            );
        }
    }
}
